File: client/src/state/commands.ts
import { create } from "zustand";
import type { SlashCommand } from "@/lib/types";
import { listCommands } from "@/api/endpoints";

interface CommandsState {
  commands: SlashCommand[];
  /** Fetched once per session on the shell's mount. A failure leaves the list
   * empty, which just means no `/` menu — never a blocked composer. */
  fetch: () => Promise<void>;
}

export const useCommandsStore = create<CommandsState>()((set) => ({
  commands: [],
  fetch: async () => {
    try {
      set({ commands: await listCommands() });
    } catch {
      set({ commands: [] });
    }
  },
}));

File: app/lib/state/commands.dart
import 'package:campfire/models/command.dart';
import 'package:campfire/state/api.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';

/// The slash commands this server's bots answer to, fetched once per session.
///
/// A failure is not an error state on purpose: an empty list just means no `/`
/// menu, which is exactly what a deployment without bots looks like. Same
/// contract as `state/commands.ts`.
class CommandsNotifier extends AsyncNotifier<List<SlashCommand>> {
  @override
  Future<List<SlashCommand>> build() async {
    try {
      return await ref.read(apiProvider).listCommands();
    } on Object {
      return const [];
    }
  }
}

final commandsProvider =
    AsyncNotifierProvider<CommandsNotifier, List<SlashCommand>>(CommandsNotifier.new);
